libs/exec/exec.go

```diff
@@ -0,0 +1,249 @@
+package exec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	osexec "os/exec"
+)
+
+type Command interface {
+	// Wait for command to terminate. It must have been previously started.
+	Wait() error
+
+	// StdinPipe returns a pipe that will be connected to the command's standard input when the command starts.
+	Stdout() io.ReadCloser
+
+	// StderrPipe returns a pipe that will be connected to the command's standard error when the command starts.
+	Stderr() io.ReadCloser
+}
+
+type command struct {
+	done   chan bool
+	err    chan error
+	stdout io.ReadCloser
+	stderr io.ReadCloser
+}
+
+func (c *command) Wait() error {
+	select {
+	case <-c.done:
+		return nil
+	case err := <-c.err:
+		return err
+	}
+}
+
+func (c *command) Stdout() io.ReadCloser {
+	return c.stdout
+}
+
+func (c *command) Stderr() io.ReadCloser {
+	return c.stderr
+}
+
+type Executor struct {
+	interpreter interpreter
+	dir         string
+}
+
+func NewCommandExecutor(dir string) (*Executor, error) {
+	interpreter, err := findInterpreter()
+	if err != nil {
+		return nil, err
+	}
+	return &Executor{
+		interpreter: interpreter,
+		dir:         dir,
+	}, nil
+}
+
+func (e *Executor) StartCommand(ctx context.Context, command string) (Command, error) {
+	e.interpreter.prepare(command)
```

Comment:
This is the only stateful part; if someone uses the same executor to run 2 commands concurrently, then the first one to terminate will clean up the file created by the last one to start. If this is pulled out of the interpreter state and onto the stack then it looks safe. Can be done as a follow-up, as long as we document it is not safe to run 2 commands concurrently.